### ISSUE TITLE
population_template: Affine shear drift

### ISSUE THREAD

**Issue by Lestropie** [MRtrix3/mrtrix3#827]
Currently `population_template` runs one linear stage (either rigid or affine), then one non-linear stage.

Images below are from generating a template from 8 T1 scans of the same volunteer within 2 days, so images should be near-identical - but the subject was instructed to re-position their head in between scans, so there's a fair bit of rotation & translation.

First image shows in cyan the initial template when rigid registration is selected for the linear stage, and in red the initial template when affine registration is selected:

![screenshot0000](https://cloud.githubusercontent.com/assets/5637955/20468770/2620f010-afee-11e6-9d95-e16c2f251a3a.png)

Second image shows the results of a hack-modified `population_template` script: in cyan is the _first_ initial template where _rigid_ registration was used, and in red is the _second_ initial template where _affine_ registration was used, having been initialised using the transforms from the earlier rigid registration:

![screenshot0001](https://cloud.githubusercontent.com/assets/5637955/20468820/a8ac646a-afee-11e6-968c-6014db730405.png)

Clearly when affine registration is used without being initialised using transforms from rigid registration, there are erroneous shear components with non-zero mean being introduced while correcting for translations/rotations, distorting the template. If rigid registration is first used, there's no residual image gradients present to drive those shear components.

Possible solutions:

* Hardwire `population_template` to use rigid then affine registration by default. Probably bring the script interface more in line with `mrregister`, with separate settings for all three stages of registration.

* At each iteration, decompose the linear transformations and de-mean all translation, rotation and shear components across the population.

**Comment by draffelt** [comment]
I like the idea of using rigid first. Although I guess this is less important for inter-subject registration than intra-subject registration. Max, what are you thoughts? Is the rigid registration as robust as the affine?

**Comment by thijsdhollander** [comment]
I agree it should always make sense to do rigid first, and then affine.  (so happy with such a change!)

...but in general, this is always an issue when all images can move: there's no fixed frame of reference any more that keeps things in place.  Even when doing rigid registration first, there's still nothing that constrains the subsequent affine registration to not start drifting, also in the non-rigid dimensions (e.g. shear).  It's of course much less likely to happen when initialised with the outcome from a robust rigid registration.

Even our non-linear symmetric registration suffers "in principle" from this effect: there's nothing that discourages joint drifting of both images in the linear dimensions.  Even an elastic regulariser (i.e. Gaussian smoothing of the current warp after each iteration) does not fight this (at all): the images can jointly start (globally) shearing happily.  In practice, this effect is typically limited, but in principle it's there!

**Comment by maxpietsch** [comment]
Happy to include an optional/ additional rigid stage in `population_template`.

> Is the rigid registration as robust as the affine

The rigid registration should in principle be as or more robust than the affine. The only problem is that the gradient descent update step is an affine update which is then projected to be orthonormal. This mismatch between gradient estimation and parameter update makes it slow to converge. I have an idea how to prevent this, just need to finish implementing it...

For the `population_template`, I would offer a rigid only option and a rigid_affine, where we limit the number of rigid iterations to let's say 50 and then continue with the affine registration.

@Lestropie How much rotation are we talking about? I would use `-affine_init_rotation search` for large rotations between subjects or `-affine_init_rotation moments` for longitudinal studies. 

> At each iteration, decompose the linear transformations and de-mean all translation, rotation and shear components across the population.

I try to prevent "drifting" in `population_template` by factoring out the average affine transformation at each iteration level. That should in practice accelerate convergence to a stable common space. Not sure if convergence for rigid only registration would be faster if the population template is run on different subjects.

**Comment by maxpietsch** [comment]
I've had a go at the proposed interface change: 446a294c7895a328d 

    Options for the linear registration
    
      -linear_type
         Specifiy the linear registration mode. Options are "affine" (perform affine
         registration only), "rigid" (perform rigid registration only, this should
         be used for intra-subject registration in longitudinal analysis) and
         "rigid_affine" (perform rigid registration followed by affine registration,
         default).
    
      -linear_no_pause
         Do not pause the script if a linear registration seems implausible
    
      -linear_estimator LINEAR_ESTIMATOR
         Choose estimator for intensity difference metric. Valid choices are: l1
         (least absolute: |x|), l2 (ordinary least squares), lp (least powers:
         |x|^1.2), Default: l2
    
      -initial_alignment
         Method of alignment to form the initial template. Options are "mass"
         (default), "geometric" and "none".
    
      -rigid_scale RIGID_SCALE
         Specifiy the multi-resolution pyramid used to build the rigid template, in
         the form of a list of scale factors (default: 0.3,0.4,0.6,0.8,1.0,1.0).
         This and affine_scale implicitly  define the number of template levels
    
      -rigid_lmax RIGID_LMAX
         Specifiy the lmax used for rigid registration for each scale factor, in the
         form of a list of integers (default: 2,2,2,4,4,4). The list must be the
         same length as the linear_scale factor list
    
      -rigid_niter RIGID_NITER
         Specifiy the number of registration iterations used within each level
         before updating the template, in the form of a list of integers (default:50
         for each scale). This must be a single number or a list of same length as
         the linear_scale factor list
    
      -affine_scale AFFINE_SCALE
         Specifiy the multi-resolution pyramid used to build the affine template, in
         the form of a list of scale factors (default: 0.3,0.4,0.6,0.8,1.0,1.0).
         This and rigid_scale implicitly define the number of template levels
    
      -affine_lmax AFFINE_LMAX
         Specifiy the lmax used for affine registration for each scale factor, in
         the form of a list of integers (default: 2,2,2,4,4,4). The list must be the
         same length as the linear_scale factor list
    
      -affine_niter AFFINE_NITER
         Specifiy the number of registration iterations used within each level
         before updating the template, in the form of a list of integers
         (default:500 for each scale). This must be a single number or a list of
         same length as the linear_scale factor list

Internally, I use lists that hold parameters for all linear steps irrespective of rigid or affine to keep variable clutter down. Haven't tested the changes yet but if you're happy with this, then I'll do that.

**Comment by Lestropie** [comment]
One thing @draffelt and I had discussed was the possibility of generating _just_ a linear template; i.e. skipping the non-linear stage. So in that case, you'd really want the interface to mimic `mrregister`, i.e. have registration mode options:

```
rigid,affine,nonlinear,rigid_affine,rigid_nonlinear,affine_nonlinear,rigid_affine_nonlinear
```

, with the last being the default.

> I try to prevent "drifting" in `population_template` by factoring out the average affine transformation at each iteration level. 

I presume you mean [this section](https://github.com/MRtrix3/mrtrix3/blob/master/scripts/population_template#L427-L436)? My understanding of transformations & matrix math is not in the top three in this discussion, but is this precisely mathematically equivalent to decomposing each subject transformation into translations / rotations / scales / shears, de-meaning each parameter across the population (maybe accounting for axis permutations in the case of very large rotations), then re-composing each subject transformation? I'd expect the latter to address the fundamental problem; question is (a) whether I'm right, (b) whether it's worth the effort.

**Comment by maxpietsch** [comment]
> rigid,affine,nonlinear,rigid_affine,rigid_nonlinear,affine_nonlinear,rigid_affine_nonlinear

Sounds good to me. I'll implement it when I have a second.

> I presume you mean this section

Yes, that's the section.

> precisely mathematically equivalent to decomposing each subject transformation into translations / rotations / scales / shears
One problem is that the decomposition is not unique. [Here](https://github.com/MRtrix3/mrtrix3/blob/master/cmd/transformcalc.cpp#L72) is how I do it. Even if it were, another problem is that his decomposition is not very useful. For instance, translation and rotation (around the centre of the coordinate system) are not independent. The same holds for shear and rotation with scale. So you have to factor out the full affine (or full rigid) transformation in one step. The way to do that is to calculate the [mean](https://github.com/MRtrix3/mrtrix3/blob/2bf78387e7b5dca6f2ea0adf0783b87b5f489bb1/lib/math/average_space.cpp#L24) of the affine transformations (in the affine manifold) and factor it out of the individual transformation matrices by applying the inverse mean affine transformation.

So, I think we already do the best mathematically speaking. But using a rigid registration first and then the affine makes sense as it should solve the problem you showed above.

**Comment by maxpietsch** [comment]
I've changed the script and tested a few parameter combinations. Testing is welcome! 

Also, we'll need to figure out sensible default values for each registration configuration. The default of rigid + affine seems a bit overkill but is sensible if only one of them is specified:

    $ population_template csd template.mif -mask_dir dwi_mask
    population_template: Generating a population-average template from 20 input images
    population_template: SH Series detected, performing FOD registration
    population_template: initial alignment of images: mass
    population_template: linear registration stages:
    population_template: (00) rigid     scale: 0.3000, niter: 100, lmax: 2
    population_template: (01) rigid     scale: 0.4000, niter: 100, lmax: 2
    population_template: (02) rigid     scale: 0.6000, niter: 100, lmax: 2
    population_template: (03) rigid     scale: 0.8000, niter: 100, lmax: 4
    population_template: (04) rigid     scale: 1.0000, niter: 100, lmax: 4
    population_template: (05) rigid     scale: 1.0000, niter: 100, lmax: 4
    population_template: (06) affine    scale: 0.3000, niter: 500, lmax: 2
    population_template: (07) affine    scale: 0.4000, niter: 500, lmax: 2
    population_template: (08) affine    scale: 0.6000, niter: 500, lmax: 2
    population_template: (09) affine    scale: 0.8000, niter: 500, lmax: 4
    population_template: (10) affine    scale: 1.0000, niter: 500, lmax: 4
    population_template: (11) affine    scale: 1.0000, niter: 500, lmax: 4
    population_template: nonlinear registration stages:
    population_template: (00) nonlinear scale: 0.3000, niter: 5, lmax: 2
    population_template: (01) nonlinear scale: 0.4000, niter: 5, lmax: 2
    population_template: (02) nonlinear scale: 0.5000, niter: 5, lmax: 2
    population_template: (03) nonlinear scale: 0.6000, niter: 5, lmax: 2
    population_template: (04) nonlinear scale: 0.7000, niter: 5, lmax: 2
    population_template: (05) nonlinear scale: 0.8000, niter: 5, lmax: 2
    population_template: (06) nonlinear scale: 0.9000, niter: 5, lmax: 2
    population_template: (07) nonlinear scale: 1.0000, niter: 5, lmax: 2
    population_template: (08) nonlinear scale: 1.0000, niter: 5, lmax: 4
    population_template: (09) nonlinear scale: 1.0000, niter: 5, lmax: 4
    population_template: (10) nonlinear scale: 1.0000, niter: 5, lmax: 4
    population_template: (11) nonlinear scale: 1.0000, niter: 5, lmax: 4
    population_template: (12) nonlinear scale: 1.0000, niter: 5, lmax: 4
    population_template: (13) nonlinear scale: 1.0000, niter: 5, lmax: 4
    population_template: (14) nonlinear scale: 1.0000, niter: 5, lmax: 4
    population_template: (15) nonlinear scale: 1.0000, niter: 5, lmax: 4
    ...

**Comment by maxpietsch** [comment]
Tested the script on some data where the affine only option failed and it worked beautifully :) 

I also implemented the convergence check that uses double exponential smoothing of the control point trajectories: registration stops if the predicted movement (slope in the middle plot below) of all control points for the next gradient descent iteration is below 1/500 * voxel_size.

I tested it on 5 pairs of anatomical, FA, and FOD images of typical resolution using rigid an affine registration. Registration is faster or equally fast at comparable results for the majority of cases.

Examples where it is good:

![image](https://cloud.githubusercontent.com/assets/10046944/20833797/6d811590-b889-11e6-829f-b4a3dd03092d.png)
(trajectories are for the non-convergence check option only)

![image](https://cloud.githubusercontent.com/assets/10046944/20833987/6eb76b0c-b88a-11e6-8caf-9527f8983bf9.png)

Example where earlier stopping in lower resolution stage causes more iterations in higher iteration stages:
![image](https://cloud.githubusercontent.com/assets/10046944/20834015/963de3b8-b88a-11e6-89a4-ed80d724c4de.png)
